src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java

```diff
@@ -0,0 +1,950 @@
+        }
+    }
+
+    static class CategorizedSSTables
+    {
+        public final List<SSTableStats> data;
+        public final List<SSTableStats> expiredTombstones;
+        public final List<SSTableStats> nearExpiryTombstones;
+        public final List<SSTableStats> latestTombstones;
+        public final OverallStats stats;
+
+        public CategorizedSSTables(List<SSTableStats> data, List<SSTableStats> expiredTombstones, List<SSTableStats> nearExpiryTombstones, List<SSTableStats> latestTombstones, OverallStats stats)
+        {
+            this.data = data;
+            this.expiredTombstones = expiredTombstones;
+            this.nearExpiryTombstones = nearExpiryTombstones;
+            this.latestTombstones = latestTombstones;
+            this.stats = stats;
+        }
+
+        @Override
+        public String toString()
+        {
+            return "CategorizedSSTables{" +
+                    "data=" + idString(data) +
+                    ", expiredTombstones=" + idString(expiredTombstones) +
+                    ", nearExpiryTombstones=" + idString(nearExpiryTombstones) +
+                    ", latestTombstones=" + idString(latestTombstones) +
+                    ", stats=" + stats +
+                    '}';
+        }
+    }
+
+    CategorizedSSTables categorizeSStables(Iterable<SSTableReader> sstables, int gcBefore)
+    {
+        int gcGraceSecondsMs = Math.max(1, cfs.metadata.params.gcGraceSeconds) * 1000;
+        long nowMs = System.currentTimeMillis();
+        long currentGcWindowMs = toWindowMs(nowMs, gcGraceSecondsMs);
+
+        // tombstone sstables that have expired and now can be compacted with data sstables to free up space.
+        List<SSTableStats> compactableTombstones = new ArrayList<>(Iterables.size(sstables));
+        // tombstones sstables that are only created recently and cannot be deleted. These will be meregd together to keep the overlap between sstables low.
+        List<SSTableStats> latestTombstones = new ArrayList<>();
+        // tombstone sstables that are about to be expired. They most likely will have 0 overlap. But if not (due to streaming), then merge it.
+        List<SSTableStats> nearExpiryTombstones = new ArrayList<>();
+        // data sstables
+        List<SSTableStats> dataSSTables = new ArrayList<>();
+
+        OverallStats overallStats = new OverallStats();
+        sstables.forEach(s -> {
+            SSTableStats stats = getOrCompute(s);
+            collect(overallStats, stats);
+            if (s.getSSTableLevel() == SSTable.TOMBSTONE_SSTABLE_LVL)
+            {
+                // If gcGraceSeconds > 0, i.e. tombstones will linger in sstables for some time and if they are created recently,
+                // mark them as latest so that they are not compacted against data sstables just yet.
+                if (cfs.metadata.params.gcGraceSeconds > 0 && toWindowMs(s.getSSTableMetadata().maxLocalDeletionTime * 1000, gcGraceSecondsMs) >= currentGcWindowMs)
```

Comment:
I'm not clear why we're considering both gcBefore and the current time (currentGcWindowMs). Shouldn't it be entirely based on gcBefore? My understanding is that the reason for gcBefore param is that all methods have unified view of what the cutoff point is for gc, instead of everyone calling now() and getting inconsistent results.

IMO from what I understand of the categorization, we should just figure out the window corresponding to gcBefore, then if window of maxLocalDeletion time of table is:
> gcWindow => latest
= gcWindow => nearExpiry (will become expirable in next window)
< gcWindow => compactable

Reply:
We cannot use gcBefore instead of currentGcWindowMs.
If I do that, there will be 2 categories then. 
<------------------->|<--------->
          expired                     latest
And I intend to merge all the latest tombstones together. In which case the merged file can never get expired. Note that the tombstone sstable expires when its latest delete timestamp < gcBefore. But merging latest sstables may not allow that to happen.

So, i went with a fix window and created 3 categories.
<------------------->|<--->|<-------->
        expired              near expiry   latest
The latest window is fixed, and we will merge whatever sstables lie in that window. As soon as that window passes, they will fall into "near expiry" window and soon into expired window.

And about the first if check, the check gcGraceSeconds > 0 means that if gcGraceSeconds = 0 then the sstable will never be latest, they are expired as soon as they are created. To be honest the same check should be there for "near expiry" also.